ftw-time doc (#196)
sharetribe/flex-docs: src/docs/background/20-time-based-template/index.md

---
title: Time-based template
slug: time-based-template
updated: 2019-10-21
category: background
ingress:
  This article introduces the time-based template and what's different compared
  to the day-based FTW.
published: true
---

Time-based template adds support for time-based bookings. By default, template
uses a transaction process
[preauth-unit-time-booking](https://github.com/sharetribe/flex-example-processes/tree/master/preauth-unit-time-booking)
which is a unit-based process and has value `time` in `create-booking` booking
action.

> **Note:** We have removed `BookingDatesForm` and `EditListingAvailabilityForm`
> components form the template as they are heavily related to the day-based
> process. If you have started with the original FTW and you are planning to
> move to the time-based template you should check if you have done any
> customization to these components.
>
> You should also note that `DateRangeFilter` is not in use in the time-based
> template because the API does not support querying time-based availability.

## Handling the timezones

> With time-based bookings timezones play a larger role than with night- and

Reviewer: I think this should be "time zone".
https://en.wikipedia.org/wiki/Time_zone

Check all the other usages too (except when referring to moment-timezone lib)

day-based ones. This is the case especially if your marketplace operates on
multiple timezones and some listings could be booked from a different timezone
as the one in which the listing is located.

FTW-time introduces a new library `moment-timezone` for handling the timezones
when operating with dates. You need to use the moment-timezone whenever you
create new date objects. E.g. on ListingPage the start and end of the month
should have timezone so that the moment-timezone library can handle e.g daylight
savings.

The timezone of the listing will be saved to the listing's
[availability plan](https://www.sharetribe.com/api-reference/#ownlisting-availability-plan).

> For that, we have a new component `FieldTimeZoneSelect` where the user can

Reviewer: I would change the 'user' to 'provider' to enforce our vocabulary.

choose the timezone of the listing. By default, the timezone will be the

> timezone of user's browser.

Reviewer: "By default, the timezone will be the timezone of user's browser."
=> "By default, the component will select browser's time zone."

## Booking the listing

> In the time-based template, the minimum duration of the booking is 1 hour and

Reviewer: ```suggestion
In the time-based template, the default minimum duration of the booking is 1 hour and
```

the available booking times are starting on sharp hours (e.g. 9:00, 10:00...).

> As the minimum length of the timeslot is 5 minutes duration and starting times

Reviewer: ```suggestion
Since the minimum duration of the timeslot is 5 minutes and starting times
```

can be customized to support different use-cases. If you want to change the
duration or timing of the timeslot you should start from
[util/dates.js](https://github.com/sharetribe/ftw-time/blob/master/src/util/dates.js)
file. You should edit at least the `findNextBoundary`, `getSharpHours` and

> `calculateQuantityFromHours` functions.

Reviewer: ```
Since the minimum duration of the timeslot is 5 minutes and starting times
can be customized to support different use-cases, you might want to change the
duration or timing of your timeslot-units. Start this process from
[util/dates.js](https://github.com/sharetribe/ftw-time/blob/master/src/util/dates.js)
file. You should edit at least the `findNextBoundary`, `getSharpHours` and
`calculateQuantityFromHours` functions.
```

All the dates are in the listing's timezone. There are new components
`BookingTimeForm` and `FieldDateAndTimeInput` which handle querying the
available timeslots and picking the booking start and end times. Also, all the
dates in the email templates should also be in the listing's timezone. Listing's
availability plan and the timezone are available in transaction email context

![Booking panel](bookingPanel.png 'BookingPanel')

## Managing availability

In the EditListingWizard you can save and edit the listing's default schedule.

> Each day can have one or more available timeslots. You can also have days when

Reviewer: "timeslot" is confusing here. First of all, I think it should be "time slot" or "time-slot", but it would be better if we avoid it altogether since these entries are not referring to the same time-slots as our API endpoint for retrieving time slots.

"Each day can have several entries which mark bookable periods of time. So, you can add holes to your business hours with entries like 09 - 12 and 14 - 18."
